cmd/cstor-pool-mgmt/volumereplica/volumereplica.go

```diff
@@ -669,3 +671,172 @@ func GetAndUpdateReplicaID(cvr *apis.CStorVolumeReplica) error {
+	dsName := PoolNameFromCVR(cvr) + "/" + volName
+
+	snapList, err := GetSnapshotList(dsName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get the list of snapshots")
+	}
+
+	// Add/Delete a snapshot if it doesn't exist in CVR but exists in ZFS
+	err = addOrDeleteSnapshotListInfo(cvr, snapList)
+	if err != nil {
+		return errors.Wrapf(err, "failed to add snapshot list info")
+	}
+
+	// If CVR is in rebuilding go and get snapshot information
+	// from other replicas
+	//if cvr.Status.Phase == apis.CVRStatusRebuilding {
+	//	err = getAndUpdatePendingSnapshotsList(clientset, cvr)
+	//}
+	return nil
+}
+
+//func getAndUpdatePendingSnapshotsList(
+//	clientset clientset.Interface, cvr *apis.CStorVolumeReplica) error {
+//	peerCVRList, err := getPeerReplicas(clientset, cvr)
+//	if err != nil {
+//		return errors.Wrapf(err, "failed to get peer CVRs info of volume %s", volName)
+//	}
+// TODO: Perform following steps
+// 1. Get the snapshot info from peer replicas
+// 2. If snapshot info doesn't exist under CVR.Status.Snapshots then
+// add it under CVR.Status.PendingSnapshots
+//}
+
+// addOrDeleteSnapshotListInfo adds/deletes the snapshots in CVR
+// It performs below steps:
+// 1. Add snapshot if it doesn't exist in CVR but exist on ZFS.
+// 2. Delete snapshot if exist in CVR but not in ZFS.
+// 3. Delete pending snapshots in CVR if exist in ZFS.
+// NOTE: Below function will get the snapshot info from ZFS
+func addOrDeleteSnapshotListInfo(
+	cvr *apis.CStorVolumeReplica,
+	currentSnapList map[string]string) error {
+	var err error
+	var snapInfo apis.CStorSnapshotInfo
+	volName := cvr.GetLabels()[string(apis.PersistentVolumeCPK)]
+	dsName := PoolNameFromCVR(cvr) + "/" + volName
+
+	if cvr.Status.Snapshots == nil {
+		cvr.Status.Snapshots = map[string]apis.CStorSnapshotInfo{}
+	}
+
+	// Add snapshot if it doesn't exist in CVR but exist on ZFS<Paste>
+	for snapName, _ := range currentSnapList {
+		// If snapshot doesn't exist in CVR.Status.Snapshots then
+		// get the snapshot info from zfs and Update info in CVR.Status.Snapshots
+		if _, ok := cvr.Status.Snapshots[snapName]; !ok {
+			snapInfo, err = getSnapshotInfo(dsName, snapName)
+			if err != nil {
+				return errors.Wrapf(
+					err,
+					"failed to get the properties of snapshot %s@%s",
+					dsName, snapName)
+			}
+			cvr.Status.Snapshots[snapName] = snapInfo
+		}
+	}
+
+	// Delete snapshot if exist in CVR but not in ZFS
+	for snapName, _ := range cvr.Status.Snapshots {
+		if _, ok := currentSnapList[snapName]; !ok {
+			delete(cvr.Status.Snapshots, snapName)
+		}
+	}
+
+	// Delete pending snapshots in CVR if exist in ZFS
+	for snapName, _ := range cvr.Status.PendingSnapshots {
+		if _, ok := currentSnapList[snapName]; ok {
+			delete(cvr.Status.PendingSnapshots, snapName)
+		}
+	}
+	return nil
+}
+
+// GetSnapshotList get the list of snapshots by executing
+// command: `zfs listsnap <dataset_name>` and returns error if
```

Comment:
paste the sample output for both success and error cases..
also, the error codes

Reply:
I will add

Reply:
>  the error codes
Why error codes are required in comments?

Reply:
Can we get this list of error codes somewhere? I am aware of only 11 in this scenario.